File: week2/stream_client.py
"""Manual test client for Exercise 1's SSE endpoint. Run the server first:
    uvicorn app:app --reload
then in another terminal:
    python stream_client.py "What is a race condition?"
"""

import json
import sys
import time

import httpx


def stream_question(question, base_url="http://127.0.0.1:8000"):
    start = time.perf_counter()
    event = "message"
    with httpx.Client(timeout=None) as client:
        with client.stream("GET", f"{base_url}/ask/stream", params={"question": question}) as response:
            response.raise_for_status()
            for line in response.iter_lines():
                if line == "":
                    event = "message"
                    continue
                if line.startswith("event:"):
                    event = line[len("event:") :].strip()
                    continue
                if not line.startswith("data:"):
                    continue

                data = line[len("data:") :].strip()
                if event == "start":
                    print(f"[start] {data}\n")
                elif event == "done":
                    print(f"\n\n[done] {data}")
                elif event == "error":
                    print(f"\n\n[error] {data}")
                else:
                    print(_extract_token(data), end="", flush=True)

    print(f"\nClient-side wall time: {round(time.perf_counter() - start, 2)}s")


def _extract_token(data_json):
    try:
        return json.loads(data_json).get("token", "")
    except json.JSONDecodeError:
        return ""


if __name__ == "__main__":
    question_text = " ".join(sys.argv[1:]) or "Explain what Server-Sent Events are in three sentences."
    stream_question(question_text)

File: week2-js/src/streamClient.ts


async function streamQuestion(question: string, baseUrl = "http://127.0.0.1:8000"): Promise<void> {
  const start = performance.now();
  let event = "message";

  const response = await fetch(`${baseUrl}/ask/stream?${new URLSearchParams({ question })}`);
  if (!response.ok || !response.body) {
    throw new Error(`request failed: ${response.status} ${response.statusText}`);
  }

  const reader = response.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";

  while (true) {
    const { done, value } = await reader.read();
    if (done) break;

    buffer += decoder.decode(value, { stream: true });
    const lines = buffer.split("\n");
    buffer = lines.pop() ?? ""; // keep the last (possibly incomplete) line for next chunk

    for (const rawLine of lines) {
      const line = rawLine.replace(/\r$/, ""); // fetch doesn't strip \r before \n like httpx does
      if (line === "") {
        event = "message";
        continue;
      }
      if (line.startsWith("event:")) {
        event = line.slice("event:".length).trim();
        continue;
      }
      if (!line.startsWith("data:")) {
        continue;
      }

      const data = line.slice("data:".length).trim();
      if (event === "start") {
        process.stdout.write(`[start] ${data}\n\n`);
      } else if (event === "done") {
        process.stdout.write(`\n\n[done] ${data}\n`);
      } else if (event === "error") {
        process.stdout.write(`\n\n[error] ${data}\n`);
      } else {
        process.stdout.write(extractToken(data));
      }
    }
  }

  const elapsed = Math.round((performance.now() - start) / 10) / 100;
  process.stdout.write(`\nClient-side wall time: ${elapsed}s\n`);
}

function extractToken(dataJson: string): string {
  try {
    return JSON.parse(dataJson).token ?? "";
  } catch {
    return "";
  }
}

const questionText = process.argv.slice(2).join(" ") || "Explain what Server-Sent Events are in three sentences.";
streamQuestion(questionText).catch((error) => {
  console.error(error);
  process.exit(1);
});
